backprojection.cpp

```diff
@@ -251,6 +255,90 @@ class BackprojectionGenerator : public Halide::Generator<BackprojectionGenerator
+                        .vectorize(sample_vi)
+                        .parallel(sample_vo);
+            win_sample.w.update(0)
+                        .split(sample, sample_vo, sample_vi, vectorsize, TailStrategy::GuardWithIf)
+                        .vectorize(sample_vi)
+                        .parallel(sample_vo);
+            win_pulse.taylor.compute_root()
+                            .vectorize(pulse, vectorsize)
+                            .parallel(pulse, blocksize);
+            win_pulse.w.compute_root()
+                       .split(pulse, pulse_vo, pulse_vi, vectorsize)
+                       .vectorize(pulse_vi)
+                       .parallel(pulse_vo);
+            win_pulse.w.update(0)
+                       .split(pulse, pulse_vo, pulse_vi, vectorsize, TailStrategy::GuardWithIf)
+                       .vectorize(pulse_vi)
+                       .parallel(pulse_vo);
+            win.compute_root();
+            filt.compute_root();
+            phs_filt.inner.compute_root()
+                          .bound(c, 0, 2).unroll(c)
+                          .vectorize(sample, vectorsize)
+                          .parallel(pulse, blocksize);
+            phs_pad.inner.compute_root()
+                         .bound(c, 0, 2).unroll(c)
+                         .vectorize(sample, vectorsize)
+                         .parallel(pulse, blocksize);
+            fftsh.inner.compute_root()
+                       .bound(c, 0, 2).unroll(c)
+                       .vectorize(sample, vectorsize)
+                       .parallel(pulse, blocksize);
+            dft.inner.compute_root().parallel(pulse);
+            Q.inner.compute_at(img.inner, rnpulses.x)
+                   .store_at(img.inner, rnpulses.x)
+                   .store_in(MemoryType::Heap)
+                   .split(sample, sample_vo, sample_vi, vectorsize)
+                   .vectorize(sample_vi)
+                   .parallel(pulse);
+            norm_r0.compute_root();
+            rr0.compute_inline();
+            norm_rr0.compute_inline();
+            dr_i.compute_inline();
+            Q_hat.inner.compute_inline();
+            img.inner.compute_root()
+                     .gpu_tile(pixel, x_vo, x_vi, blocksize_gpu_tile);
+            img.inner.update(0)
+                     .reorder(c, pixel, rnpulses.x)
+                     .gpu_tile(pixel, x_vo, x_vi, blocksize_gpu_tile);
+            fimg.inner.compute_root()
+                      .reorder(c, x, y)
+                      .gpu_blocks(y)
+                      .gpu_tile(x, x_vo, x_vi, blocksize_gpu_tile);
+            output_img.compute_root()
+                      .bound(c, 0, 2)
+                      .unroll(c)
+                      .vectorize(x, vectorsize)
+                      .parallel(y);
+#if defined(WITH_DISTRIBUTE)
+            if (sched == Schedule::GPUSplitDistributed) {
+                output_img.distribute(y).send_to(0);
```

Comment:
Appears to require the newer distributed Halide, which sarbp is not updated for yet.

Reply:
Memory usage of the normal `cuda` schedule on a single node:
```
calling _kernel_fimg_s0_y___block_id_x with gridsize 2048x1x1, blocksize 16x1x1, shmem 0 and parameters:
name               | dtype   | shape           | bytes    
----------------------------------------------------------
k_r                | float32 | [1800]          | 7680     
norm_r0            | float32 | [1999]          | 8192     
pos                | float32 | [3, 1999]       | 24576    
fimg               | float64 | [2, 2048, 2048] | 67108864 
r_in               | float64 | [4194304, 3]    | 100663296
Q                  | float64 | [2, 4096, 1999] | 134217728
norm_rr0 temporary | float64 | [31984, 2048]   | 536870912
total GPU data size: 838901248
```

Memory usage of this `cuda_split` schedule on the same problem size:
```
calling _kernel_img_s1_pixel_x_vo___block_id_x__1 with gridsize 262144x1x1, blocksize 16x1x1, shmem 0 and parameters:
name    | dtype   | shape        | bytes    
--------------------------------------------
k_r     | float32 | [1800]       | 7680     
norm_r0 | float32 | [1999]       | 8192     
pos     | float32 | [3, 1999]    | 24576    
Q[i]    | float64 | [2, 4096, 1] | 65536    
img     | float64 | [2, 4194304] | 67108864 
r_in    | float64 | [4194304, 3] | 100663296
total GPU data size: 167878144
```

Reply:
Good catch, fixed.